File: algorithms/2_medium/S/00560_subarray_sum_equals_k/hashmap.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        auto hashmap = unordered_map<int, int>();
        hashmap[0] = 1;
        int sum = 0;
        int cnt = 0;

        for (const int num : nums) {
            sum += num;

            auto it = hashmap.find(sum - k);
            if (it != hashmap.end()) {
                cnt += it->second;
            }

            hashmap[sum] += 1;
        }

        return cnt;
    }
};

File: algorithms/2_medium/S/00560_subarray_sum_equals_k/hashmap.rs
use std::collections::HashMap;

impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut hashmap: HashMap<i32, i32> = HashMap::with_capacity(nums.len() + 1);
        hashmap.insert(0, 1);
        let mut sum = 0_i32;
        let mut cnt = 0_i32;

        for num in nums {
            sum += num;
            if let Some(freq) = hashmap.get(&(sum - k)) {
                cnt += *freq;
            }

            *hashmap.entry(sum).or_insert(0) += 1;
        }

        cnt
    }
}
